### PR TITLE
Add a `Place::is_indirect` method to determine whether a `Place` contains a `Deref` projection

### PR DESCRIPTION
**Pull request by ecstatic-morse**
Working on #63860 requires tracking some property about each local. This requires differentiating `Place`s like `x` and `x.field[index]` from ones like `*x` and `*x.field`, since the first two will always access the same region of memory as `x` while the latter two may access any region of memory. This functionality is duplicated in various places across the compiler. This PR adds a helper method to `Place` which determines whether that `Place` has a `Deref` projection at any point and changes some existing code to use the new method.

I've not converted `qualify_consts.rs` to use the new method, since it's not a trivial conversion and it will get replaced anyway by #63860. There may be other potential uses besides the two I change in this PR.

r? @oli-obk

### CODE REVIEW

File: src/librustc_mir/borrow_check/path_utils.rs [hunk 1]
```diff
@@ -133,20 +133,13 @@ pub(super) fn is_active<'tcx>(
...
-                return false;
-            }
-        }
+        // Reborrow of already borrowed data is ignored
+        // Any errors will be caught on the initial borrow
+        _ if place.is_indirect() => false,
 
-        true
-    })
+        PlaceBase::Local(_) => true,
```
Comment (oli-obk) [inline src/librustc_mir/borrow_check/path_utils.rs]: What do you think about changing the `true` to `!place.is_indirect()`?

**Review by oli-obk — COMMENTED** [review]
From the changes in this PR it looks to me like all use cases could use the `find_local` function instead?

File: src/librustc_mir/borrow_check/path_utils.rs (hunk 1 above)
Comment (ecstatic-morse) [reply src/librustc_mir/borrow_check/path_utils.rs]: That works better.

File: src/librustc_mir/dataflow/impls/borrowed_locals.rs [hunk 2]
```diff
@@ -93,19 +93,15 @@ struct BorrowedLocalsVisitor<'gk> {
...
-            Some(*local)
-        } else {
-            None
-        }
-    })
+    if let PlaceBase::Local(local) = place.base {
+        Some(local)
+    } else {
+        None
+    }
```
Comment (ecstatic-morse) [inline src/librustc_mir/dataflow/impls/borrowed_locals.rs]:
```suggestion
        match place.base {
        PlaceBase::Local(local) if !place.is_indirect() => Some(local),
        _ => None,
    }
```
A more streamlined version of this...

**Review by spastorino — APPROVED** [review]

**Comment by spastorino** [comment]
Nice work and looks good to me. I'm not sure I like the name of the function but can't think of a better one right now. Any other ideas?, @oli-obk do you consider the fn name good enough? any better ideas?.

**Comment by oli-obk** [comment]
@bors r+

Nope, no better ideas here

**Pull request merged**